### PR TITLE
CEP-15: Convert AccordStateCache cache from write-through to write-back

### PR DESCRIPTION
**Pull request by iamaleksey**
patch by Aleksey Yeschenko; reviewed by Blake Eggleston for CASSANDRA-18563

### CODE REVIEW

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java [hunk 1]
```diff
@@ -15,547 +15,442 @@
...
+     * for longer to maximise the chances of hitting system tables fewer times (or not at all).
+     */
+    private void maybeEvictSomeNodes()
     {
         if (bytesCached <= maxSizeInBytes)
             return;
 
-        Node<?, ?> current = tail;
-        while (current != null && bytesCached > maxSizeInBytes)
+        Iterator<AccordCachingState<?, ?>> iter = this.iterator();
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/accord/AccordStateCache.java:117]: Can we avoid allocating an iterator here? This is called all the time. Maybe we could add a peek method and manually iterate through the linked list?

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java [hunk 2]
```diff
@@ -15,547 +15,442 @@
...
+                    // schedule a save to disk, keep linked and in the cache map
+                    Instance<?, ?, ?> instance = instanceForNode(node);
+                    node.save(saveExecutor, instance.saveFunction);
+                    maybeUpdateSize(node, instance.heapEstimator);
+                    break;
+                case SAVING:
+                    // skip over until completes to LOADED or FAILED_TO_SAVE
+                    break;
+                case FAILED_TO_SAVE:
+                    // permanently unlink, but keep in the map; consider panicking instead when this happens
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/accord/AccordStateCache.java]: Can you add a TODO for this?

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java [hunk 3]
```diff
@@ -15,547 +15,442 @@
...
+                    evict(node);
+                    break;
+                case MODIFIED:
+                    // schedule a save to disk, keep linked and in the cache map
+                    Instance<?, ?, ?> instance = instanceForNode(node);
+                    node.save(saveExecutor, instance.saveFunction);
+                    maybeUpdateSize(node, instance.heapEstimator);
+                    break;
+                case SAVING:
+                    // skip over until completes to LOADED or FAILED_TO_SAVE
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/accord/AccordStateCache.java:142]: We should preserve the TODO about skipping unselectable nodes

**Review by bdeggleston — CHANGES_REQUESTED** [review]
A few minor things. This is a really nice refinement of some of the caching stuff

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java (hunk 1 above)
Comment (iamaleksey) [reply src/java/org/apache/cassandra/service/accord/AccordStateCache.java:117]: Introduced iterator caching to `IntrusiveLinkedList`, like Agrona does in all of its single-threaded collections, here: https://github.com/apache/cassandra-accord/pull/51

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java (hunk 2 above)
Comment (iamaleksey) [reply src/java/org/apache/cassandra/service/accord/AccordStateCache.java]: Sure, converted to a TODO.

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java (hunk 3 above)
Comment (iamaleksey) [reply src/java/org/apache/cassandra/service/accord/AccordStateCache.java:142]: I think it's sufficiently preserved, above, just above the switch statement? Unless you mean something else.

Comment (bdeggleston) [reply src/java/org/apache/cassandra/service/accord/AccordStateCache.java:142]: heh, oh right, missed it there :)

**Review by bdeggleston — APPROVED** [review]
+1

File: src/java/org/apache/cassandra/service/accord/AccordStateCache.java (hunk 1 above)
Comment (iamaleksey) [reply src/java/org/apache/cassandra/service/accord/AccordStateCache.java:117]: This ended up being unnecessary (see comments on the linked cassandra-accord PR).

**Comment by iamaleksey** [comment]
Merged manually, cheers.

**Pull request closed without merging**